fcrepo-kernel-impl/src/main/java/org/fcrepo/kernel/impl/TransactionImpl.java

```diff
@@ -30,40 +37,69 @@
  */
 public class TransactionImpl implements Transaction {
 
+    private final static Logger log = LoggerFactory.getLogger(TransactionImpl.class);
+
     final String id;
 
+    final TransactionManagerImpl txManager;
+
+    final Duration DEFAULT_TIMEOUT = ofMinutes(3);
+
     boolean shortLived = true;
 
-    TransactionImpl(final String id) {
+    Instant expiration;
+
+    boolean expired = false;
+
+    boolean rolledback = false;
+
+    boolean commited = false;
+
+    protected TransactionImpl(final String id, final TransactionManagerImpl txManager) {
         if (id == null || id.isEmpty()) {
             throw new IllegalArgumentException("Transaction id should not be empty!");
         }
         this.id = id;
+        this.txManager = txManager;
+        this.expiration = Instant.now().plus(timeout());
     }
 
     @Override
     public void commit() {
-        // Prepare Persistence Transactions
-        // Commit Persistence Transactions
+        failIfExpired();
+        failIfRolledback();
+        if(this.commited) {
+            return;
+        }
+        try {
+            log.debug("Commiting transaction {}", id);
+            this.getPersistentSession().commit();
+            this.commited = true;
+        } catch (final PersistentStorageException ex) {
+            throw new RepositoryRuntimeException("failed to commit transaction " + id, ex);
+        }
     }
 
     @Override
     public void rollback() {
-        // Rollback Persistence Transactions
-
-        // Delete Transaction from TransactionManager state?
+        failIfCommited();
+        if(this.rolledback) {
+            return;
+        }
+        try {
+            log.debug("Rolling back transaction {}", id);
+            this.getPersistentSession().rollback();
+            this.rolledback = true;
```

Comment:
whether or not rollback succeeds, I think we still want to mark the transaction as closed/rolled back to prevent further changes. So maybe move this up a line?